ampd/src/starknet/verifier.rs

```diff
@@ -0,0 +1,348 @@
+use axelar_wasm_std::voting::Vote;
+use cosmwasm_std::HexBinary;
+use starknet_core::types::Felt;
+use starknet_types::events::contract_call::ContractCallEvent;
+use starknet_types::events::signers_rotated::SignersRotatedEvent;
+
+use crate::handlers::starknet_verify_msg::Message;
+use crate::handlers::starknet_verify_verifier_set::VerifierSetConfirmation;
+
+/// Attempts to fetch the tx provided in `axl_msg.tx_id`.
+/// If successful, extracts and parses the ContractCall event
+/// and compares it to the message from the relayer (via PollStarted event).
+/// Also checks if the source_gateway_address with which
+/// the voting verifier has been instantiated is the same address from
+/// which the ContractCall event is coming.
+pub fn verify_msg(
+    starknet_event: &ContractCallEvent,
+    msg: &Message,
+    source_gateway_address: &str,
+) -> Vote {
+    if *starknet_event == *msg && starknet_event.from_contract_addr == source_gateway_address {
+        Vote::SucceededOnChain
+    } else {
+        Vote::NotFound
+    }
+}
+
+impl PartialEq<Message> for ContractCallEvent {
+    fn eq(&self, axl_msg: &Message) -> bool {
+        Felt::from(axl_msg.source_address.clone()) == self.source_address
+            && axl_msg.destination_chain == self.destination_chain
+            && axl_msg.destination_address == self.destination_address
+            && axl_msg.payload_hash == self.payload_hash
+    }
+}
+
+pub fn verify_verifier_set(
+    event: &SignersRotatedEvent,
+    confirmation: &VerifierSetConfirmation,
+    source_gateway_address: &str,
+) -> Vote {
+    if event.signers.nonce != [0_u8; 32]
+        && event == confirmation
+        && event.from_address == source_gateway_address
+    {
+        Vote::SucceededOnChain
+    } else {
+        Vote::NotFound
+    }
+}
+
+impl PartialEq<VerifierSetConfirmation> for SignersRotatedEvent {
+    fn eq(&self, confirmation: &VerifierSetConfirmation) -> bool {
+        let expected = &confirmation.verifier_set;
+
+        // Convert and sort expected signers
+        let mut expected_signers = expected
+            .signers
+            .values()
+            .map(|signer| {
+                if let multisig::key::PublicKey::Ecdsa(pubkey) = &signer.pub_key {
+                    (pubkey.clone(), signer.weight.u128())
+                } else {
+                    // Skip non-ECDSA keys
+                    (HexBinary::from_hex("").unwrap(), 0)
```

Comment:
Is this not an expected case? If so I think you should return an error from `map`, instead of this empty data.